### PR TITLE
feat: set release of arbitrary clients

### PR DESCRIPTION
**Pull request by savudani8**
- `CurrentClientRelease` is changed from a value store to a map of type `(Vec<u8>,  ClientRelease<T::Hash>)`
- No migration is necessary, since no release has been set on-chain
- The `PendingClientRelease` field and its associated migration have been removed. Using just the `CurrentClientRelease` would only delay a client auto-update by one block plus the binary download time, which is negligible and worth the complexity reduction

### CODE REVIEW

**Review by sander2 — CHANGES_REQUESTED** [review]
> The PendingClientRelease field and its associated migration have been removed. Using just the CurrentClientRelease would only delay a client auto-update by one block plus the binary download time, which is negligible and worth the complexity reduction

I think you're missing the main goal of `PendingClientRelease`, which is to be able to batch `[authorize_upgrade, set_pending_client_release]` in a single governance proposal, and have the the CurrentRelease automatically change automatically at the exact time the upgrade is actually executed. So I think we should keep it for this reason. The side benefit that you are talking about, which is that the client can download the upgrade in advance, is imo indeed unnecessary complexity on the client side, and I am fine with not implementing that.

File: crates/vault-registry/src/types.rs [hunk 1]
```diff
@@ -121,84 +121,97 @@ pub type DefaultVaultId<T> = VaultId<<T as frame_system::Config>::AccountId, Cur
...
-            0
-        };
-        T::DbWeight::get().reads_writes(1, writes)
+    /// For each pending client release, set the current release to that.
+    /// The pending release entry is removed.
+    pub fn try_upgrade_current_client_releases<T: Config>() -> Weight {
+        let keys = crate::PendingClientRelease::<T>::iter_keys().collect::<Vec<_>>();
+        let mut writes: Weight = 0;
+        keys.iter().for_each(|key| {
+            if let Some(release) = crate::PendingClientRelease::<T>::take(key) {
```
Comment (gregdhill) [inline crates/vault-registry/src/types.rs:135]: I'm wondering if we should `drain` the `PendingClientRelease` map instead?

File: crates/vault-registry/src/lib.rs [hunk 2]
```diff
@@ -428,32 +428,42 @@ pub mod pallet {
...
-        pub fn set_current_client_release(origin: OriginFor<T>, uri: Vec<u8>, code_hash: T::Hash) -> DispatchResult {
+        pub fn set_current_client_release(
+            origin: OriginFor<T>,
+            client_name: Vec<u8>,
+            release: ClientRelease<T::Hash>,
+        ) -> DispatchResult {
             ensure_root(origin)?;
-            let release = ClientRelease { uri, code_hash };
-            CurrentClientRelease::<T>::put(release.clone());
+            CurrentClientRelease::<T>::insert(client_name, release.clone());
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:445]: Nit: `CurrentClientRelease` should be `CurrentClientReleases`

File: crates/vault-registry/src/lib.rs [hunk 3]
```diff
@@ -428,32 +428,42 @@ pub mod pallet {
...
-        pub fn set_pending_client_release(origin: OriginFor<T>, uri: Vec<u8>, code_hash: T::Hash) -> DispatchResult {
+        pub fn set_pending_client_release(
+            origin: OriginFor<T>,
+            client_name: Vec<u8>,
+            release: ClientRelease<T::Hash>,
+        ) -> DispatchResult {
             ensure_root(origin)?;
-            let release = ClientRelease { uri, code_hash };
-            PendingClientRelease::<T>::put(Some(release.clone()));
+            PendingClientRelease::<T>::insert(client_name, release.clone());
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:466]: Nit: `PendingClientRelease` should be `PendingClientReleases`

**Review by gregdhill — COMMENTED** [review]
Left a few nits and one question, but otherwise LGTM. I do also wonder if it is worth moving this logic to a separate crate however, I realize it doesn't do a whole lot but this pallet is already quite big, it's not distinctly related to the core functionality here and I can foresee related changes we may want to group together.

**Review by gregdhill — APPROVED** [review]

**Pull request merged**